### DISCUSSION TITLE
Custom geometry

### DISCUSSION THREAD

**Discussion by stromhaug-tommy** [GeometryGym/GeometryGymIFC#88] [Q&A] [answered]
Hi, in a previous version i was using the basin.cs example to generate tesselated geometry. This example does not work in the latest version. This is what i have tried :(The file loads but not showing any geometry). Does it exist a way to "convert" cad-step geometry to IFC- geometry?

            db = new DatabaseIfc(ModelView.Ifc4DesignTransfer);
            IfcBuilding building = new IfcBuilding(db, "IfcBuilding") { };
            IfcProject project = new IfcProject(building, "IfcProject", IfcUnitAssignment.Length.Millimetre) { };

            IfcBuildingStorey storey = new IfcBuildingStorey(building, "Ground Floor", 0);

            LIST<IfcRepresentationMap> rm = new LIST<IfcRepresentationMap>();
            List<IfcFace> facesShell = new List<IfcFace>();

            IfcCartesianPoint[] vertices = new IfcCartesianPoint[8];
            vertices[0] = new IfcCartesianPoint(db, 0, 0, 0);
            vertices[1] = new IfcCartesianPoint(db, 0, 0, 1);
            vertices[2] = new IfcCartesianPoint(db, 0, 1, 0);
            vertices[3] = new IfcCartesianPoint(db, 0, 1, 1);
            vertices[4] = new IfcCartesianPoint(db, 1, 0, 0);
            vertices[5] = new IfcCartesianPoint(db, 1, 0, 1);
            vertices[6] = new IfcCartesianPoint(db, 1, 1, 0);
            vertices[7] = new IfcCartesianPoint(db, 1, 1, 1);

            IfcPolyLoop pLoop1 = new IfcPolyLoop(new IfcCartesianPoint[] { vertices[0], vertices[2], vertices[6], vertices[4] });
            IfcPolyLoop pLoop2 = new IfcPolyLoop(new IfcCartesianPoint[] { vertices[1], vertices[5], vertices[7], vertices[3] });
            IfcPolyLoop pLoop3 = new IfcPolyLoop(new IfcCartesianPoint[] { vertices[0], vertices[4], vertices[5], vertices[1] });
            IfcPolyLoop pLoop4 = new IfcPolyLoop(new IfcCartesianPoint[] { vertices[2], vertices[3], vertices[7], vertices[6] });
            IfcPolyLoop pLoop5 = new IfcPolyLoop(new IfcCartesianPoint[] { vertices[0], vertices[1], vertices[3], vertices[2] });
            IfcPolyLoop pLoop6 = new IfcPolyLoop(new IfcCartesianPoint[] { vertices[4], vertices[6], vertices[7], vertices[5] });

            IfcFaceOuterBound oBound = new IfcFaceOuterBound(pLoop1, true);
            IfcFace face = new IfcFace(oBound);
            facesShell.Add(face);

            IfcFaceOuterBound oBound2 = new IfcFaceOuterBound(pLoop2, true);
            IfcFace face2 = new IfcFace(oBound2);
            facesShell.Add(face2);

            IfcFaceOuterBound oBound3 = new IfcFaceOuterBound(pLoop3, true);
            IfcFace face3 = new IfcFace(oBound3);
            facesShell.Add(face3);

            IfcFaceOuterBound oBound4 = new IfcFaceOuterBound(pLoop4, true);
            IfcFace face4 = new IfcFace(oBound);
            facesShell.Add(face4);

            IfcFaceOuterBound oBound5 = new IfcFaceOuterBound(pLoop5, true);
            IfcFace face5 = new IfcFace(oBound5);
            facesShell.Add(face5);

            IfcFaceOuterBound oBound6 = new IfcFaceOuterBound(pLoop6, true);
            IfcFace face6 = new IfcFace(oBound6);
            facesShell.Add(face6);

            IfcClosedShell cs = new IfcClosedShell(facesShell);

            IfcFacetedBrep k = new IfcFacetedBrep(cs);

            IfcShapeRepresentation cubeShape = new IfcShapeRepresentation(k);

            IfcRepresentationMap representationMap = new IfcRepresentationMap(db.Factory.Origin as IfcAxis2Placement, cubeShape);

            cubeShape.RepresentationType = "Brep";

            rm.Add(representationMap);
            
            IfcBuildingElementProxyType cubeProxy = new IfcBuildingElementProxyType(db, "test", IfcBuildingElementProxyTypeEnum.ELEMENT);
            cubeProxy.Name = "Custom Cube";
            cubeProxy.Description = "A cube created using IfcFacetedBrep and IfcShapeRepresentation";
            cubeProxy.RepresentationMaps = rm;

            db.Context.AddDeclared(cubeProxy);
            IfcElement element = cubeProxy.GenerateMappedItemElement(building, new IfcCartesianTransformationOperator3D(db));
            db.WriteFile(@"D:\temptest.ifc");

**Comment by jmirtsch** [comment, marked as answer]
Sorry, probably I renamed something and it's not so obvious as to what I renamed it to.  
I changed this line: 
IfcRepresentationMap representationMap = new IfcRepresentationMap(db.Factory.Origin2dPlace, cubeShape);

Also, which viewer or software do you check the resultant IFC file in?  I checked solibri, it didn't zoom to an object 1mm in size, so I also changed this line for testing. 
	IfcProject project = new IfcProject(building, "IfcProject", IfcUnitAssignment.Length.Metre) { };

Let me know if this doesn't help.  I believe faceted brep is deprecated in IFC4, although lots of implementations still use it.
Also, design transfer view hasn't advanced so it might not be well supported in viewers.
I've done some closed source work on converting CAD step geometry to IFC.  A lot of it is very similar but not everywhere, but advanced breps are not well supported in IFC which might be more common in STEP.  Not sure if other toolkits such as IfcOpenShell might have done more on this in an open source project.

**Reply by stromhaug-tommy** [reply]
Now it worked! Thank you!

**Reply by stromhaug-tommy** [reply]
Just one more question. What should I use instead of faceted brep for tesselated geometry ?

**Reply by jmirtsch** [reply]
IfcPolygonalFaceSet or IfcTriangulatedFaceSet were introduced in IFC4 and are more efficient ways to serialize tessellated Geometry.
https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcPolygonalFaceSet.htm
